### PR TITLE
Changes to pointer serialization

### LINKED ISSUE #3520: Errors Load function for StreamSerializer object
Hi to everyone,

If in https://github.com/KratosMultiphysics/Kratos/blob/master/kratos/tests/test_model.py in the function "test_file_serialization"  I modify/add the lines

        # ######## here we pickle the serializer
        # try:
        #     import cpickle as pickle  # Use cPickle on Python 2.7
        # except ImportError:
        #     import pickle

        # #pickle dataserialized_data
        # pickled_data = pickle.dumps(serializer, 2) #second argument is the protocol and is NECESSARY (according to pybind11 docs)

        # #overwrite the old serializer with the unpickled one
        # serializer = pickle.loads(pickled_data)

        loaded_model = KratosMultiphysics.Model()
        serializer.Load("ModelSerialization",loaded_model)
        self.assertTrue(loaded_model["Main"].HasNodalSolutionStepVariable(KratosMultiphysics.TEMPERATURE))
        self.assertTrue(loaded_model["Other"].HasNodalSolutionStepVariable(KratosMultiphysics.PRESSURE))

        self.assertTrue(loaded_model.HasModelPart("Main.Inlets"))
        self.assertTrue(loaded_model.HasModelPart("Main.Temp"))
        self.assertTrue(1 in loaded_model["Main"].Nodes)
        self.assertTrue(1 in loaded_model["Other"].Nodes)

        del(loaded_model)

        loaded_model_again = KratosMultiphysics.Model()
        serializer.Load("ModelSerialization",loaded_model_again)

I am not able to deserialize a second time the StreaSerializer object. I get the following error:
    serializer.Load("ModelSerialization",loaded_model_again)
MemoryError: std::bad_alloc

For the parameters I get a similar problem when I deserealize a second time. In https://github.com/KratosMultiphysics/Kratos/blob/master/kratos/tests/test_kratos_parameters.py in the function "test_stream_serialization" I modify/add the following lines
       
        # ######## here we pickle the serializer
        # try:
        #     import cickle as pickle  # Use cPickle on Python 2.7
        # except ImportError:
        #     import pickle

        # #pickle dataserialized_data
        # pickled_data = pickle.dumps(serializer, 2) #second argument is the protocol and is NECESSARY (according to pybind11 docs)
        # del(serializer)

        #unpickle data - note that here i override "serialized_data"
        # serializer = pickle.loads(pickled_data)

        loaded_parameters = Parameters()
        serializer.Load("ParametersSerialization",loaded_parameters)

        loaded_parameters_again = Parameters()
        serializer.Load("ParametersSerialization",loaded_parameters_again)
to deserialize a second time the StreamSerializer Kratos object, and I get the following error:
The document is empty. offset of the error from the beginning of the string = 0

I attach the two files I modifed.

[Serializer_errors.zip](https://github.com/KratosMultiphysics/Kratos/files/2624430/Serializer_errors.zip)

### PR DESCRIPTION
**Pull request by roigcarlo**
As we commented, until we implement the fix with the pointer table.
closes #3520

### CODE REVIEW

**Comment by roigcarlo** [comment]
So, at the end just commenting the code for global_pointer_vectors gives some problems with a test I don't want to disable.

My alternative is to initialize the mLoadedPointer and mSavePointer in the serializer before the actual save and load.

@pooyan-dadvand This should technically be the same as the trick we did with the global pointers, but with the benefit of being inside the serializer itself.

I am aware that this could potentially give problems if some class has a constructor/initializer that is being call by the serializer that for some reason is dependent on some value which has been delayed to load. To be honest I can't come up with any examples but just in case I would like to ask you guys if you can test or send me a tests of a complex serializer example (maybe a restart init or something like it)

Anyway to be discussed tomorrow

File: kratos/sources/serializer.cpp [hunk 1]
```diff
@@ -49,14 +49,15 @@ void Serializer::load(std::string const & rTag, ModelPart*& pValue)
...
                 KRATOS_ERROR_IF(!pValue) << "an already constructed modelpart must be passed to load a ModelPart" <<std::endl;
-
-                load(rTag, *pValue);
             }
             else if(pointer_type == SP_DERIVED_CLASS_POINTER)
             {
                 KRATOS_ERROR << "should not find SP_DERIVED_CLASS_POINTER for ModelPart load" << std::endl;
             }
+
+            // Load the pointer adress before loading the content
```
Comment (roigcarlo) [inline kratos/sources/serializer.cpp]: Typo. I am aware.

**Comment by pooyan-dadvand** [comment]
> @pooyan-dadvand This should technically be the same as the trick we did with the global pointers, but with the benefit of being inside the serializer itself.

My idea was to do this in `Serializer `not in `GlobalPointer` as you did here

**Review by pooyan-dadvand — APPROVED** [review]
I'm glad that this works

**Comment by roigcarlo** [comment]
I will merge and keep an eye opened in case something happens :+1:

**Pull request merged**